Add Channel constructor keyword argument to `@spawn` new Task in parallel

NHDaly opened the pull request:
Adds a `spawn=` keyword argument to `Channel()`, which  causes the new `Task` created for the provided `func` to be scheduled on any thread, in parallel, equivalent to creating the task via `@spawn`.

Defaults to false for backwards compatability.

This adds the param to the new constructor added in #32818, and updates the NEWS item.

NHDaly committed:
Add spawn arg to Channel(f), to create parallel Tasks

Add a parameter to the Channel constructor to allow the Tasks it creates
to be scheduled on multiple threads.

Examples:
```julia
    # Spawn a Task to handle each input request
    ch = Channel{String}(0, spawn=true) do ch
        handle_request(ch, request)
    end
```

NHDaly committed:
Add new spawn=true Channel param to NEWS

NHDaly commented on test/channels.jl:
```diff
@@ -76,6 +76,13 @@ end
+    # Test that the task is using the multithreaded scheduler
+    @test taskref[].sticky == false
+    @test collect(c) == [0]
+end
```
I don't love that this test is exposing an implementation detail (tests like this make it more annoying to change implementation details like that), but I don't know of a better way to test that the task is created using the multithreaded scheduler?

NHDaly commented:
@JeffBezanson This should be the last of my Channel constructor PRs. 😄 Thanks for the help reviewing them!

JeffBezanson commented on NEWS.md:
```diff
@@ -20,8 +20,10 @@ Multi-threading changes
 * New experimental `Threads.@spawn` macro that runs a task on any available thread ([#32600]).
-* Simplified the `Channel` constructor, which is now easier to read and more idiomatic julia.
-  The old constructor (which used kwargs) is still available, but use is discouraged ([#30855], [#32818]).
+* New `Channel(f::Function)` constructor param (`spawn=true`) to schedule created Task on
```
I think this can remain two items. Can also consistently refer to them as "keyword arguments". (In hindsight I wish we'd always called them "named arguments" but we'd have to change that in a lot of places)

JeffBezanson commented on base/channels.jl:
```diff
@@ -65,6 +65,9 @@ Create a new task from `func`, bind it to a new channel of type
 the keyword argument `taskref`.
 
+If `spawn = true`, the Task created for `func` may be scheduled on another thread
+in parallel, equivalent to creating a task via `Threads.@spawn`.
```
```suggestion
in parallel, equivalent to creating a task via [`Threads.@spawn`](@ref).
```

JeffBezanson commented on base/channels.jl:
```diff
@@ -105,11 +108,12 @@ true
 !!! compat "Julia 1.3"
-  This constructor was added in Julia 1.3. Earlier versions of Julia used kwargs
-  to set `size` and `T`, but those constructors are deprecated.
+  The `spawn=` parameter was added in Julia 1.3. This constructor was added in Julia 1.3.
```
```suggestion
  The `spawn` keyword argument was added in Julia 1.3. This constructor was added in Julia 1.3.
```

JeffBezanson commented on base/channels.jl:
````diff
@@ -120,12 +124,16 @@ julia> String(collect(chnl))
 ```
 """
-function Channel{T}(func::Function, size=0; taskref=nothing) where T
+function Channel{T}(func::Function, size=0; taskref=nothing, spawn=false) where T
````
Unfortunately the deprecated `Channel(func::Function; ctype, csize, taskref)` definition overwrites the `Channel(::Function)` case, so `spawn` can't be used with it. I guess it's simplest to add `spawn` to that definition.

NHDaly replied:
Ah shoot, you're right. :'(

I had previously added `spawn=` to the deprecated constructor when i originally wrote this a few weeks ago, but when i was cleaning it up now I was confused by that and removed it. I thought this would be a good carrot to bring people over to the new constructor syntax in order to use the `spawn=true` argument...


Hmm, is it a crazy/bad idea to add the `spawn` argument to the deprecated constructor and then also add a runtime check that the user doesn't set `spawn=true` _and_ any of the old kwargs, and if they do, throw an ArgumentError? (or fake a MethodError)?

Something like:
```julia
function Channel(func::Function; ctype=nothing, csize=nothing, taskref=nothing, spawn=nothing)
    if (ctype !== nothing || csize !== nothing) && spawn !== nothing
        throw(ArgumentError("Cannot set `spawn=` in the deprecated constructor `Channel(f; ctype=Any, csize=0)`. Please use `Channel{Any}(f, size=0; taskref=nothing, spawn=false)` instead!"))
    end
    # Set default values
    ctype === nothing && ctype=Any
    csize === nothing && csize=0
    spawn === nothing && spawn=false
    return Channel{ctype}(func, csize; taskref=taskref)
end
```

This would allow us to bring the users along to the new constructor with a "soft deprecation" of the old method: "if you want to use the new feature, you need to use the new method", in exchange for a small runtime cost. (And I think I've arranged the short-circuiting correctly above so that the correct case `Channel(spawn=true) do ch; ... end` is the fastest.)

What do you think? Is that overkill?

NHDaly replied:
Okay i'm doing this.

NHDaly committed:
Apply suggestions from code review

- improve docstrings for Channel constructors

Co-Authored-By: Jeff Bezanson <jeff.bezanson@gmail.com>

NHDaly committed:
NEWS.md: Split the Channel constructor changes back into two sep bullets

NHDaly committed:
Improve `compat` docstring

NHDaly committed:
Fix `Channel(::Function, spawn=true)` by adding spawn= to the kwarg-only constructor

Adds a manual check that users don't use the `spawn=` keyword argument
with the deprecated keyword arguments `ctype=` or `csize=`.
